src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,262 @@
+
+(defgeneric into! (function tensor)
+  (:documentation "Map indices of TENSOR by replacing the value with the output of FUNCTION on the index at the element.")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (apply function dims) tensor dims)))
+    tensor))
+
+(defgeneric foreach (function tensor)
+  (:documentation "Call FUNCTION with each element of TENSOR")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (funcall function (apply #'tref tensor dims))))
+    tensor))
+
+(defgeneric map-to (function source target)
+  (:documentation "Map elements of SOURCE by replacing the corresponding element of TARGET the output of FUNCTION on the source element")
+  (:method ((function function) (source abstract-tensor) (target abstract-tensor))
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert (equalp (shape source) (shape target))
+             () "Incompatible shapes. Cannot map tensor of shape ~a to tensor of shape ~a."
+             (shape source) (shape target))
+     nil)
+    (map-indexes
+     (shape source)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (funcall function (apply #'tref source dims)) target dims)))))
+
+(defgeneric map (function tensor)
+  (:documentation "Map elements of TENSOR, storing the output of FUNCTION on the element into the corresponding element of a new tensor")
+  (:method ((function function) (tensor abstract-tensor))
+    (map! function (deep-copy-tensor tensor))))
+
+(defgeneric into (function tensor)
+  (:documentation "Map indices of TENSOR, storing the output of FUNCTION on the index into the corresponding element of a new tensor
+
+If LAYOUT is specified then traverse TENSOR in the specified order (column major or row major).")
+  (:method ((function function) (tensor abstract-tensor))
+    (into! function (deep-copy-tensor tensor))))
+
+(defgeneric sum (tensor)
+  (:documentation "Get the sum of the elements of TENSOR")
+  (:method ((tensor abstract-tensor))
+    (let ((sum 0))
+      (foreach (lambda (x) (incf sum x)) tensor)
+      sum)))
+
+(defgeneric scale (tensor factor)
+  (:documentation "Scale TENSOR by FACTOR, returning a new tensor of the same type as TENSOR")
+  (:method ((tensor abstract-tensor) (factor number))
+    (map! (lambda (x) (* x factor)) (deep-copy-tensor tensor))))
+
+(defgeneric scale! (tensor factor)
+  (:documentation "Scale TENSOR by FACTOR, storing back into the tensor")
+  (:method ((tensor abstract-tensor) (factor number))
+    (map! (lambda (x) (* x factor)) tensor)))
+
+(defgeneric slice (tensor from to)
+  (:documentation "Slice a tensor from FROM to TO, returning a new tensor with the contained elements")
+  (:method ((tensor abstract-tensor) from to)
+    (declare (type sequence from to))
+    (policy-cond:policy-if
+     (< speed safety)
+     (progn
```

Comment:
An `if` with a `progn` form inside is the tell-tale sign that you ought to be using `cond`.